server/src/main/java/com/linecorp/centraldogma/server/command/StandaloneCommandExecutor.java

```diff
@@ -195,10 +250,68 @@ protected void doStop(@Nullable Runnable onReleaseLeadership) {
     }
 
     private CompletableFuture<Revision> push(PushCommand c) {
+        if (c.projectName().equals(INTERNAL_PROJ) || c.repositoryName().equals(Project.REPO_DOGMA) ||
+            !writeQuotaEnabled()) {
+            return push0(c);
+        }
+
+        final RateLimiter rateLimiter =
+                writeRateLimiters.get(rateLimiterKey(c.projectName(), c.repositoryName()));
+        if (rateLimiter != null) {
+            return tryPush(c, rateLimiter);
+        }
+
+        return getRateLimiter(c.projectName(), c.repositoryName()).thenCompose(limiter -> tryPush(c, limiter));
+    }
+
+    private CompletableFuture<Revision> tryPush(PushCommand c, @Nullable RateLimiter rateLimiter) {
+        if (rateLimiter == null || rateLimiter == UNLIMITED || rateLimiter.tryAcquire()) {
+            return push0(c);
+        } else {
+            return CompletableFutures.exceptionallyCompletedFuture(
+                    new TooManyRequestsException("commits", c.executionPath(), rateLimiter.getRate()));
+        }
+    }
+
+    private CompletableFuture<Revision> push0(PushCommand c) {
         return repo(c).commit(c.baseRevision(), c.timestamp(),
                               c.author(), c.summary(), c.detail(), c.markup(), c.changes());
     }
 
+    private CompletableFuture<RateLimiter> getRateLimiter(String projectName, String repoName) {
+        return metadataService.getRepo(projectName, repoName).thenApply(meta -> {
+            setWriteQuota(projectName, repoName, meta.writeQuota());
+            return writeRateLimiters.get(rateLimiterKey(projectName, repoName));
+        });
+    }
+
+    @Override
+    public final void setWriteQuota(String projectName, String repoName, @Nullable QuotaConfig writeQuota) {
+        final double permitsForRepo = writeQuota != null ? writeQuota.permitsPerSecond() : 0;
+        final double permitsPerSecond = permitsForRepo != 0 ? permitsForRepo : this.permitsPerSecond;
```

Comment:
What happen if `this.permitsPerSecond` is `-1`?

Reply:
If `this.permitsPerSecond` is `-1`, this method should not be called.
This method is called by `MetadataService.updateWriteQuota()` and `StandaloneCommandExecutor.getRateLimiter()`.

Anyway, it would better return with no-op, if `this.permitsPerSecond` is `-1`.